src/AI/JakeMan.java

```diff
@@ -846,29 +1107,122 @@ else if( um == copter ) // Consider downgrading a tank to an inf to get a copter
     return builds;
   }
 
-  /**
-   * Sort units by funds amount in descending order.
-   */
-  private static class UnitModelFundsComparator implements Comparator<Entry<UnitModel, Double>>
-  {
-    @Override
-    public int compare(Entry<UnitModel, Double> entry1, Entry<UnitModel, Double> entry2)
-    {
-      double diff = entry2.getValue() - entry1.getValue();
-      return (int) (diff * 10); // Multiply by 10 since we return an int, but don't want to lose the decimal-level discrimination.
-    }
-  }
-
   private void init(GameMap map)
   {
     infantry = myArmy.cos[0].getUnitModel(UnitModel.TROOP);
     allTanks = myArmy.cos[0].getAllModels(UnitModel.ASSAULT);
+    tank     = allTanks.get(0);
+    mdTank   = allTanks.get(1);
     antiAir  = myArmy.cos[0].getUnitModel(UnitModel.SURFACE_TO_AIR);
     copter   = myArmy.cos[0].getUnitModel(UnitModel.ASSAULT | UnitModel.AIR_LOW, false);
+
+    counterBuildSetup();
+
+    allCounterContestants = new HashSet<>();
+    for( UnitModel counterable : unitTypeToCounterMap.keySet() )
+    {
+      allCounterContestants.add(counterable);
+      for( CounterRatio ratio : unitTypeToCounterMap.get(counterable) )
+        allCounterContestants.add(ratio.counter);
+    }
+
     if( null == copter ) // I clearly don't understand this unit set, so just grab something to hedge
       copter = myArmy.cos[0].getUnitModel(UnitModel.AIR_TO_AIR, false);
 
     capPhase = new CapPhaseAnalyzer(map, myArmy);
   }
 
+  private void counterBuildSetup()
+  {
+    counterOrder = new ArrayList<>();
+    unitTypeToCounterMap = new HashMap<>();
+    counterTypeNegationMap = new HashMap<>(); // Note: we assume all UnitModels referenced in here are in the above
+    if( !myInfo.buildCounters )
+      return;
+
+    // tanks: no need to calc them beyond the Md clause
+    // copters: 1.5 copters or 1AA per
+    if( null != copter )
+    {
+      counterOrder.add(copter);
+      unitTypeToCounterMap.put(copter, new ArrayList<>());
+      var copterCounters = unitTypeToCounterMap.get(copter);
+      CounterRatio copterCounterCopter = new CounterRatio(copter, 200/3);
```

Comment:
Whence 200/3?

Reply:
2/3, but in percent because we int mathing.
I have to have 1.5 copters per enemy copter to consider all the copters copter-countered, so my copters count as 1/1.5 of an enemy copter. Multiply both by 2 for int-ness and it's 200/3.

Reply:
Okay, so this (and the other magic force-proportion numbers) are just SME-provided rules of thumb. Cool I guess, as long as they don't occur in more than one place.

Reply:
That's the plan. And naming these at all sanely would be... difficult.